staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1231,11 +1313,32 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 
 	closeFunc := func() {
 		// This means that we couldn't send event to that watcher.
-		// Since we don't want to block on it infinitely,
-		// we simply terminate it.
+		// Since we don't want to block on it infinitely, we simply terminate it.
+
+		// If the bookmarkAfterResourceVersionReceived hasn't been received,
+		// we can safely terminate the watcher. Because the client is waiting
+		// for this specific bookmark, and we even haven't received one.
+		if !c.wasBookmarkAfterResourceVersionReceived() {
+			klog.V(1).Infof("Forcing %v watcher close due to unresponsiveness: %v. len(c.input) = %v, len(c.result) = %v", c.objectType.String(), c.identifier, len(c.input), len(c.result))
+			metrics.TerminatedWatchersCounter.WithLabelValues(c.objectType.String()).Inc()
+			c.forget(false)
+			return
+		}
+		// We have seen a bookmark, but it hasn't been sent to the client yet.
+		// That means we should drain the input buffer which contains
+		// the bookmarkAfterResourceVersion we want. We do that to make progress
+		// as clients can re-establish a new watch with the given RV and receive
+		// further notifications.
+		if !c.wasBookmarkAfterResourceVersionSentLocked() {
+			c.forget(true)
```

Comment:
I think the metric should be updated here too - in the end we're terminating this watcher too (due to its slowness), just doing that gracefully.